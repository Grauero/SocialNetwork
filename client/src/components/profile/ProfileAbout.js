import React, { Component } from 'react';

class ProfileAbout extends Component {
  render() {
    return (
      <div>
        About
      </div>
    );
  }
}

export default ProfileAbout;
