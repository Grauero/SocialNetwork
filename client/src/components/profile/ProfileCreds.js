import React from 'react';
import propTypes from 'prop-types';
import Moment from 'react-moment';

const ProfileCreds = ({ experience, education }) => {
  const experienceItems = experience.map(exp => (
    <li key={exp._id} className="list-group-item">
      <h4>{exp.company}</h4>
      <p>
        <Moment format="YYYY/MM/DD">{exp.from}</Moment> -
        {exp.to === null ? (
          'Now'
        ) : (
          <Moment format="YYYY/MM/DD" data-to>
            {exp.to}
          </Moment>
        )}
      </p>
      <p>
        <strong>Position: </strong>
        {exp.title}
      </p>
      <p>
        {exp.location === '' ? null : (
          <span data-location>
            <strong>Location: </strong>
            {exp.location}
          </span>
        )}
      </p>
      <p>
        {exp.description === '' ? null : (
          <span data-description>
            <strong>Description: </strong>
            {exp.description}
          </span>
        )}
      </p>
    </li>
  ));
  const educationItems = education.map(edu => (
    <li key={edu._id} className="list-group-item">
      <h4>{edu.school}</h4>
      <p>
        <Moment format="YYYY/MM/DD">{edu.from}</Moment> -
        {edu.to === null ? (
          'Now'
        ) : (
          <Moment format="YYYY/MM/DD">{edu.to}</Moment>
        )}
      </p>
      <p>
        <strong>Degree: </strong>
        {edu.degree}
      </p>
      <p>
        <strong>Field Of Study: </strong>
        {edu.fieldOfStudy}
      </p>
      <p>
        {edu.description === '' ? null : (
          <span data-description>
            <strong>Description: </strong>
            {edu.description}
          </span>
        )}
      </p>
    </li>
  ));

  return (
    <div className="row">
      <div className="col-md-6">
        <h3 className="text-center text-info">Experience</h3>
        {experienceItems.length > 0 ? (
          <ul className="list-group" data-experience>
            {experienceItems}
          </ul>
        ) : (
          <p className="text-center">No experience provided</p>
        )}
      </div>

      <div className="col-md-6">
        <h3 className="text-center text-info">Education</h3>
        {educationItems.length > 0 ? (
          <ul className="list-group" data-education>
            {educationItems}
          </ul>
        ) : (
          <p className="text-center">No education provided</p>
        )}
      </div>
    </div>
  );
};

ProfileCreds.propTypes = {
  experience: propTypes.instanceOf(Array).isRequired,
  education: propTypes.instanceOf(Array).isRequired
};

export default ProfileCreds;
