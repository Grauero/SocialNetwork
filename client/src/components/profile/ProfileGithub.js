import React, { Component } from 'react';
import propTypes from 'prop-types';
import axios from 'axios';
import { Link } from 'react-router-dom';

class ProfileGithub extends Component {
  state = {
    clientId: '20db24bd8d7bd006f9bd',
    clientSecret: '0928f0e83c526cbffc989066f1e455f3fff6d92d',
    count: 5,
    sort: 'created: asc',
    repos: []
  };

  async componentDidMount() {
    const { username } = this.props;
    const { clientId, clientSecret, count, sort } = this.state;

    const res = await axios.get(
      `https://api.github.com/users/${username}/repos?per_page=${count}&sort=${sort}&client_id=${clientId}&client_secret=${clientSecret}`
    );

    this.setState({ repos: res.data });
  }

  render() {
    const { repos } = this.state;
    const repoItems = repos.map(repo => (
      <div className="card card-body mb-2" key={repo.id}>
        <div className="row">
          <div className="col-md-6">
            <h4>
              <Link to={repo.html_url} className="text-info" target="_blank">
                {repo.name}
              </Link>
            </h4>
            <p>{repo.description}</p>
          </div>
          <div className="col-md-6">
            <span className="badge badge-info mr-1" data-star>
              Stars: {repo.stargazers_count}
            </span>
            <span className="badge badge-secondary mr-1" data-watcher>
              Watchers: {repo.watchers_count}
            </span>
            <span className="badge badge-success" data-fork>
              Forks: {repo.forks_count}
            </span>
          </div>
        </div>
      </div>
    ));

    return (
      <div>
        <hr />
        <h3 className="mb-4">Latest Github repos</h3>
        {repoItems}
      </div>
    );
  }
}

ProfileGithub.propTypes = {
  username: propTypes.string.isRequired
};

export default ProfileGithub;
