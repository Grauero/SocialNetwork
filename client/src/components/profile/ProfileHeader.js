import React, { Component } from 'react';

class ProfileHeader extends Component {
  render() {
    return (
      <div>
        Header
      </div>
    );
  }
}

export default ProfileHeader;
