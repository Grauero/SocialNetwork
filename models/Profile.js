const mongoose = require('mongoose');

// schema for profile document in DB
const ProfileSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'users'
  },
  handle: {
    type: String,
    required: true,
    max: 40
  },
  company: {
    type: String
  },
  website: {
    type: String
  },
  location: {
    type: String
  },
  status: {
    type: String,
    required: true
  },
  skills: {
    type: [String],
    required: true
  },
  bio: {
    type: String
  },
  githubUserName: {
    type: String
  },
  experience: [
    {
      title: {
        type: String,
        required: true
      },
      company: {
        type: String,
        required: true
      },
      location: {
        type: String
      },
      from: {
        type: Date,
        required: true
      },
      to: {
        type: Date
      },
      current: {
        type: Boolean,
        default: false
      },
      description: {
        type: String
      }
    },
  ],
  education: [
    {
      school: {
        type: String,
        required: true
      },
      degree: {
        type: String,
        required: true
      },
      fieldOfStudy: {
        type: String,
        required: true
      },
      from: {
        type: Date,
        required: true
      },
      to: {
        type: Date
      },
      current: {
        type: Boolean,
        default: false
      },
      description: {
        type: String
      }
    },
  ],
  social: {
    youtube: {
      type: String
    },
    twitter: {
      type: String
    },
    facebook: {
      type: String
    },
    linkedin: {
      type: String
    },
    instagramm: {
      type: String
    }
  },
  date: {
    type: Date,
    default: Date.now
  }
});

const Profile = mongoose.model('profile', ProfileSchema);

module.exports = Profile;
