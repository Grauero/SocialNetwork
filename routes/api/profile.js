const express = require('express');
const passport = require('passport');

const Profile = require('../../models/Profile');
const validateProfileInput = require('../../validation/profile');
const validateExperienceInput = require('../../validation/experience');
const validateEducationInput = require('../../validation/education');

const router = express.Router();

// @route   GET api/profile
// @desc    Get current users profile
// @access  Private
router.get('/', passport.authenticate('jwt', { session: false }), (req, res) => {
  const errors = {};

  Profile.findOne({ user: req.user.id })
    .populate('user', ['name', 'avatar'])
    .then((profile) => {
      if (!profile) {
        errors.noprofile = 'There is no profile';
        return res.status(404).json(errors);
      }

      res.json(profile);
    })
    .catch(err => res.status(404).json(err));
});

// @route   GET api/profile/all
// @desc    Get all profiles
// @access  Public
router.get('/all', (req, res) => {
  const errors = {};

  Profile.find()
    .populate('user', ['name', 'avatar'])
    .then((profiles) => {
      if (!profiles) {
        errors.noProfiles = 'There are no profiles';
        res.status(404).json(errors);
      }

      res.json(profiles);
    })
    .catch(() => {
      errors.handle = 'Internal server error';
      return res.status(500).json(errors);
    });
});

// @route   GET api/profile/handle/:handle
// @desc    Get profile by handle
// @access  Public
router.get('/handle/:handle', (req, res) => {
  const errors = {};

  Profile.findOne({ handle: req.params.handle })
    .populate('user', ['name', 'avatar'])
    .then((profile) => {
      if (!profile) {
        errors.noProfile = 'Profile doesnt exist';
        res.status(404).json(errors);
      }

      res.json(profile);
    })
    .catch(() => {
      errors.handle = 'Internal server error';
      return res.status(500).json(errors);
    });
});

// @route   GET api/profile/user/:user_id
// @desc    Get profile by user id
// @access  Public
router.get('/user/:user_id', (req, res) => {
  const errors = {};

  Profile.findOne({ user: req.params.user_id })
    .populate('user', ['name', 'avatar'])
    .then((profile) => {
      if (!profile) {
        errors.noProfile = 'Profile doesnt exist';
        res.status(404).json(errors);
      }

      res.json(profile);
    })
    .catch(() => {
      errors.handle = 'Internal server error';
      return res.status(500).json(errors);
    });
});

// @route   POST api/profile
// @desc    Create or update user profile
// @access  Private
router.post('/', passport.authenticate('jwt', { session: false }), (req, res) => {
  const { errors, isValid } = validateProfileInput(req.body);

  // check validation
  if (!isValid) {
    return res.status(400).json(errors);
  }

  const profileFields = {};
  profileFields.user = req.user.id;
  if (req.body.handle) profileFields.handle = req.body.handle;
  if (req.body.company) profileFields.company = req.body.company;
  if (req.body.website) profileFields.website = req.body.website;
  if (req.body.location) profileFields.location = req.body.location;
  if (req.body.bio) profileFields.bio = req.body.bio;
  if (req.body.status) profileFields.status = req.body.status;
  if (req.body.githubUserName) profileFields.githubUserName = req.body.githubUserName;

  // Skills
  if (typeof req.body.skills !== 'undefined') {
    profileFields.skills = req.body.split(',');
  }

  // Social
  profileFields.social = {};
  if (req.body.youtube) profileFields.social.youtube = req.body.youtube;
  if (req.body.twitter) profileFields.social.twitter = req.body.twitter;
  if (req.body.facebook) profileFields.social.facebook = req.body.facebook;
  if (req.body.linkedin) profileFields.social.linkedin = req.body.linkedin;
  if (req.body.instagram) profileFields.social.instagram = req.body.instagram;

  Profile.findOne({ user: req.user.id }).then((profile) => {
    if (profile) {
      // Update
      Profile.findOneAndUpdate(
        { user: req.user.id },
        { $set: profileFields },
        { new: true }
      )
        .then(profile => res.json(profile))
        .catch(() => {
          errors.handle = 'Update error';
          return res.status(500).json(errors);
        });
    } else {
      // Create

      // Check if handle exists
      Profile.findOne({ handle: profileFields.handle }).then((profile) => {
        if (profile) {
          errors.handle = 'That handle already exists';
          res.status(400).json(errors);
        }

        // Save
        new Profile(profileFields).save()
          .then(profile => res.json(profile))
          .catch(() => {
            errors.handle = 'Creation error';
            return res.status(500).json(errors);
          });
      });
    }
  });
});

// @route   POST api/profile/experience
// @desc    Add experience to profile
// @access  Private
router.post('/experience', passport.authenticate('jwt', { session: false }), (req, res) => {
  const { errors, isValid } = validateExperienceInput(req.body);

  // check validation
  if (!isValid) {
    return res.status(400).json(errors);
  }

  Profile.findOne({ user: req.user.id })
    .then((profile) => {
      const {
        title, company, location, from, to, current, description
      } = req.body;
      const newExperience = {
        title, company, location, from, to, current, description
      };

      profile.experience.unshift(newExperience);
      profile.save()
        .then(profile => res.json(profile))
        .catch(() => {
          errors.handle = 'Internal server error';
          return res.status(500).json(errors);
        });
    })
    .catch(() => {
      errors.handle = 'Internal server error';
      return res.status(500).json(errors);
    });
});

// @route   POST api/profile/education
// @desc    Add education to profile
// @access  Private
router.post('/education', passport.authenticate('jwt', { session: false }), (req, res) => {
  const { errors, isValid } = validateEducationInput(req.body);

  // check validation
  if (!isValid) {
    return res.status(400).json(errors);
  }

  Profile.findOne({ user: req.user.id })
    .then((profile) => {
      const {
        school, degree, fieldOfStudy, from, to, current, description
      } = req.body;
      const newEducation = {
        school, degree, fieldOfStudy, from, to, current, description
      };

      profile.experience.unshift(newEducation);
      profile.save()
        .then(profile => res.json(profile))
        .catch(() => {
          errors.handle = 'Internal server error';
          return res.status(500).json(errors);
        });
    })
    .catch(() => {
      errors.handle = 'Internal server error';
      return res.status(500).json(errors);
    });
});

module.exports = router;
