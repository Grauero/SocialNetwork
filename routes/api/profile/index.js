const express = require('express');

const profile = require('./profile');
const education = require('./education');
const experience = require('./experience');
const profileHandler = require('./profileHandler');

const router = express.Router();
router.use('/', profile);
router.use('/education', education);
router.use('/experience', experience);
router.use(profileHandler);

module.exports = router;
