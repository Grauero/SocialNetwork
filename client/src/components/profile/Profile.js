import React, { Component } from 'react';
import propTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import ProfileAbout from './ProfileAbout';
import ProfileCreds from './ProfileCreds';
import ProfileGithub from './ProfileGithub';
import ProfileHeader from './ProfileHeader';
import Spinner from '../common/Spinner';
import { getProfileByHandle } from '../../store/actions/profileActions';

class Profile extends Component {
  componentDidMount() {
    if (this.props.match.params.handle) {
      this.props.getProfileByHandle(this.props.match.params.handle);
    }
  }

  render() {
    return (
      <div>
        <ProfileHeader />
        <ProfileAbout />
        <ProfileCreds />
        <ProfileGithub />
      </div>
    );
  }
}

Profile.propTypes = {
  profile: propTypes.instanceOf(Object).isRequired,
  match: propTypes.instanceOf(Object).isRequired,
  getProfileByHandle: propTypes.func.isRequired
};

const mapStateToProps = state => ({
  profile: state.profile
});

export default connect(mapStateToProps, { getProfileByHandle })(Profile);
