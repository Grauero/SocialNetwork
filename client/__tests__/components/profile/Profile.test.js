import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Profile } from '../../../src/components/profile/Profile';

const props = {
  profile: {
    profile: {
      education: [],
      experience: [],
      user: { name: 'name' },
      skills: []
    },
    loading: false
  },
  match: { params: { handle: 'handle' } },
  history: { push: jest.fn() },
  getProfileByHandle: jest.fn()
};
let component;

beforeEach(
  () =>
    (component = mount(
      <BrowserRouter>
        <Profile {...props} />
      </BrowserRouter>
    ))
);

afterEach(() => component.unmount());

it('calls getProfileByHandle when component is mount', () => {
  expect(props.getProfileByHandle).toHaveBeenCalled();
  expect(props.getProfileByHandle).toHaveBeenCalledWith(
    props.match.params.handle
  );
});

it('renders <ProfileHeader />, <ProfileAbout />, <ProfileCreds /> components if profile exists', () => {
  const profileHeader = component.find('ProfileHeader');
  const profileAbout = component.find('ProfileAbout');
  const profileCreds = component.find('ProfileCreds');

  expect(profileHeader.debug()).toBeTruthy();
  expect(profileAbout.debug()).toBeTruthy();
  expect(profileCreds.debug()).toBeTruthy();
});

it('renders <ProfileGithub /> component if user have github user name in profile', () => {
  const testProps = Object.assign({}, props, {
    profile: { profile: { ...props.profile.profile, githubUserName: 'name' } }
  });
  component = shallow(<Profile {...testProps} />);
  const githubUserName = component.find('ProfileGithub');

  expect(toJSON(githubUserName)).toMatchSnapshot();
});

it('did NOT renders <ProfileGithub /> component if user dont have github user name in profile', () => {
  const githubUserName = component.find('ProfileGithub');

  expect(githubUserName.debug()).toBeFalsy();
});

it('renders <Spinner /> component if profile doesnt exists', () => {
  const testProps = Object.assign({}, props, {
    profile: { profile: null }
  });

  component = mount(
    <BrowserRouter>
      <Profile {...testProps} />
    </BrowserRouter>
  );
  const spinner = component.find('Spinner');

  expect(toJSON(spinner)).toMatchSnapshot();
});

it('renders <Spinner /> component if profile loading', () => {
  const testProps = Object.assign({}, props, {
    profile: { ...props.profile, loading: true }
  });

  component = mount(
    <BrowserRouter>
      <Profile {...testProps} />
    </BrowserRouter>
  );
  const spinner = component.find('Spinner');

  expect(toJSON(spinner)).toMatchSnapshot();
});
