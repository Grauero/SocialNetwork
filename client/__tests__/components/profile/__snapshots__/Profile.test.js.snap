// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <ProfileGithub /> component if user have github user name in profile 1`] = `
<ProfileGithub
  username="name"
/>
`;

exports[`renders <Spinner /> component if profile doesnt exists 1`] = `
<Spinner>
  <div>
    <img
      alt="Loading..."
      src="test-file-stub"
      style={
        Object {
          "display": "block",
          "margin": "auto",
          "width": "200px",
        }
      }
    />
  </div>
</Spinner>
`;

exports[`renders <Spinner /> component if profile loading 1`] = `
<Spinner>
  <div>
    <img
      alt="Loading..."
      src="test-file-stub"
      style={
        Object {
          "display": "block",
          "margin": "auto",
          "width": "200px",
        }
      }
    />
  </div>
</Spinner>
`;
